internal/server/v1/observations/derived_series.go

```diff
@@ -0,0 +1,507 @@
+	for _, varInfo := range c.varInfoMap {
+		filteredSeriesCandidates := []*varCalculatorSeries{}
+		for _, series := range varInfo.seriesCandidates {
+			if _, ok := longestDateKeySet[series.dateKey()]; ok {
+				filteredSeriesCandidates = append(filteredSeriesCandidates, series)
+			}
+		}
+		varInfo.chosenSeries = rankVarCalculatorSeries(filteredSeriesCandidates)
+	}
+
+	return nil
+}
+
+// Recursively iterate through the AST and compute the result.
+//
+// The existing ast.Walk() doesn't work here, because in ast.Walk(), visiting children is the last
+// step of the function that visits parent, so the information in child nodes cannot be passed
+// above in the iteration.
+func (c *varCalculator) evaluateExpr(node ast.Node) (*varCalculatorSeries, error) {
+	// If a node is of type *ast.Ident, it is a leaf with a series value.
+	// Otherwise, it might be *ast.ParenExpr or *ast.BinaryExpr, so we continue recursing it to
+	// compute the series value for the subtree..
+	computeChildSeries := func(node ast.Node) (*varCalculatorSeries, error) {
+		if reflect.TypeOf(node).String() == "*ast.Ident" {
+			return c.varInfoMap[node.(*ast.Ident).Name].chosenSeries, nil
+		}
+		return c.evaluateExpr(node)
+	}
+
+	// Compute new series value of the *ast.BinaryExpr.
+	// Supported operations are: +, -, *, /.
+	evaluateBinaryExpr := func(
+		x, y *varCalculatorSeries, op token.Token) (*varCalculatorSeries, error) {
+		res := &varCalculatorSeries{points: []*pb.PointStat{}}
+
+		// Upper stream guarantees that x.points and y.points have same dates.
+		seriesLength := len(x.points)
+
+		for i := 0; i < seriesLength; i++ {
+			xVal := x.points[i].GetValue()
+			yVal := y.points[i].GetValue()
+			var val float64
+			switch op {
+			case token.ADD:
+				val = xVal + yVal
+			case token.SUB:
+				val = xVal - yVal
+			case token.MUL:
+				val = xVal * yVal
+			case token.QUO:
+				if yVal == 0 {
+					return nil, fmt.Errorf("denominator cannot be zero")
+				}
+				val = xVal / yVal
+			default:
+				return nil, fmt.Errorf("unsupported op (token) %v", op)
+			}
+			res.points = append(res.points, &pb.PointStat{
+				Date:  x.points[i].GetDate(),
+				Value: val,
+			})
+		}
+
+		return res, nil
+	}
+
+	switch t := node.(type) {
+	case *ast.BinaryExpr:
+		xSeries, err := computeChildSeries(t.X)
+		if err != nil {
+			return nil, err
+		}
+		ySeries, err := computeChildSeries(t.Y)
+		if err != nil {
+			return nil, err
+		}
+		return evaluateBinaryExpr(xSeries, ySeries, t.Op)
+	case *ast.ParenExpr:
+		return c.evaluateExpr(t.X)
+	default:
+		return nil, fmt.Errorf("unsupported ast type %T", t)
+	}
+}
+
+// This implements ast.Visitor, used by ast.Walk, which iterates AST with a depth-first order.
+// For each ast.BinaryExpr, it extracts the "Name" property of X and Y.
+// For example, for an expression "Count_Person - Count_Person_Female[m=USCensus]", it returns
+// {Count_Person, Count_Person_Female[m=USCensus]}.
+type varNameVisitor struct {
```

Comment:
maybe move all the pure AST parsing to a separate file?

Reply:
done